rust/wal3/src/lib.rs

```diff
@@ -0,0 +1,184 @@
+#![doc = include_str!("../README.md")]
+
+use serde::{Deserialize, Serialize};
+
+mod backoff;
+
+use backoff::ExponentialBackoff;
+
+/////////////////////////////////////////////// Error //////////////////////////////////////////////
+
+/// Error captures the different error conditions that can occur when interacting with the log.
+#[derive(Clone, Debug, Default)]
+pub enum Error {
+    #[default]
+    Success,
+    UninitializedLog,
+    AlreadyInitialized,
+    GarbageCollected,
+    LogContention,
+    LogFull,
+    EndOfLog,
+    Internal,
+    CorruptManifest(String),
+    CorruptCursor(String),
+    NoSuchCursor(String),
+}
+
+//////////////////////////////////////////// LogPosition ///////////////////////////////////////////
+
+/// A log position is a pair of an offset and a timestamp.  Every record has a unique log position.
+/// A LogPosition only implements equality, which checks both offset and timestamp_us.
+#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, serde::Deserialize, serde::Serialize)]
+pub struct LogPosition {
+    /// The offset field of a LogPosition is a strictly increasing timestamp.  It has no gaps and
+    /// spans [0, u64::MAX).
+    offset: u64,
+    /// The timestamp_us field of a LogPosition is a strictly increasing timestamp.  It has gaps
+    /// and corresponds to wallclock time.
+    timestamp_us: u64,
+}
+
+impl LogPosition {
+    pub fn new(offset: u64, timestamp_us: u64) -> Self {
+        LogPosition {
+            offset,
+            timestamp_us,
+        }
+    }
+
+    pub fn offset(&self) -> u64 {
+        self.offset
+    }
+
+    pub fn timestamp_us(&self) -> u64 {
+        self.timestamp_us
+    }
+
+    pub fn contains_offset(start: LogPosition, end: LogPosition, offset: u64) -> bool {
+        start.offset <= offset && offset < end.offset
+    }
+
+    pub fn contains_timestamp(start: LogPosition, end: LogPosition, timestamp: u64) -> bool {
+        start.offset <= timestamp && timestamp < end.offset
+    }
+}
+
+////////////////////////////////////////// ThrottleOptions /////////////////////////////////////////
+
+/// ThrottleOptions control admission to S3 and batch size/interval.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct ThrottleOptions {
+    /// The maximum number of bytes to batch.  Defaults to 8MB.
```

Comment:
Should we add serde defaults? We'll eventually want this to be able to be configured with our config patterns?